### PR TITLE
Add callouts for L3 plugin driver

### PR DESCRIPTION
**Pull request by tbachman**
This adds the callouts that are missing for the L3
plugin driver API.

Signed-off-by: Thomas Bachman tbachman@yahoo.com

### CODE REVIEW

File: networking_cisco/plugins/cisco/db/l3/l3_router_appliance_db.py [hunk 1]
```diff
@@ -223,6 +223,11 @@ def create_router(self, context, router):
             # Ensure create spec is compliant with any HA
             ha_spec = self._ensure_create_ha_compliant(r, router_type)
         auto_schedule, share_host = self._ensure_router_scheduling_compliant(r)
+        driver = self._get_router_type_driver(context,
+                                              router_type_id)
+        if driver:
+            router_ctxt = driver_context.RouterContext(router)
+            driver.create_router_precommit(context, router_ctxt)
```
Comment (kentwu1) [inline networking_cisco/plugins/cisco/db/l3/l3_router_appliance_db.py:230]: overall looks good to me. Some questions:
1. I wonder what's the convention to call those pre and post-commit(). In this case, only create_router_precommit() is called but not the postcommit(). In some other case its the other way around. Then in the FIP cases, both pre and post commits are called.....So I'm getting a bit confused here.

Comment (tbachman) [inline networking_cisco/plugins/cisco/db/l3/l3_router_appliance_db.py:230]: It's a good question. I only added the callouts that are needed for our APIC L3 driver. I probably should add the callouts for the other ones, even though our driver won't have implementations for them. I'll update this patch to add callouts for everything defined by the API.

**Comment by tbachman** [comment]
Updated, based on feedback.

File: networking_cisco/plugins/cisco/db/l3/l3_router_appliance_db.py [hunk 2]
```diff
@@ -233,12 +238,22 @@ def create_router(self, context, router):
...
+        router_type_id = self.get_router_type_id(context, id)
+        driver = self._get_router_type_driver(context,
+                                              router_type_id)
+        if driver:
+            router_ctxt = driver_context.RouterContext(router)
+            driver.update_router_precommit(context, router_ctxt)
         router_updated = self._update_router_no_notify(context, id, router)
         self.add_type_and_hosting_device_info(context.elevated(),
                                               router_updated)
+        driver = self._get_router_type_driver(context,
```
Comment (kentwu1) [inline networking_cisco/plugins/cisco/db/l3/l3_router_appliance_db.py]: this line looks redundant. I think u meant to call update_router_postcommit() instead.

File: networking_cisco/plugins/cisco/db/l3/l3_router_appliance_db.py [hunk 3]
```diff
@@ -414,6 +454,10 @@ def add_router_interface(self, context, router_id, interface_info):
                                                            info['port_id']))
         routers = [self.get_router(context, router_id)]
         self.add_type_and_hosting_device_info(context.elevated(), routers[0])
+        driver = self._get_router_type_driver(context,
```
Comment (kentwu1) [inline networking_cisco/plugins/cisco/db/l3/l3_router_appliance_db.py]: why do we have to query again to get the driver? Can't we just re-use the driver we got earlier when we do precommit()?

Comment (tbachman) [inline networking_cisco/plugins/cisco/db/l3/l3_router_appliance_db.py]: You're right -- will remove.

File: networking_cisco/plugins/cisco/db/l3/l3_router_appliance_db.py (hunk 2 above)
Comment (tbachman) [inline networking_cisco/plugins/cisco/db/l3/l3_router_appliance_db.py]: You're right -- I just need to delete this line (the postcommit call is made inside _update_router_no_notify). Note that I still need to get the router type driver, as the router can change during an update (i.e. the router -- and therefore router type driver -- can be different in the original vs. the ending). Will remove.

Comment (tbachman) [inline networking_cisco/plugins/cisco/db/l3/l3_router_appliance_db.py]: Actually, correction -- the router will be the same in pre- and post- for this API call.

**Comment by tbachman** [comment]
New patch set available, addressing reviewer comments.

**Pull request merged**